(ci) Update macOS Xcode to 15

ehuss opened the pull request:
This updates the macOS builders to Xcode 15. The aarch64 images will be removing Xcode 14 and 16 very soon (https://github.com/actions/runner-images/issues/10703), so we will need to make the switch to continue operating. The linked issue also documents GitHub's new policy for how they will be updating Xcode in the future. Also worth being aware of is the future plans for x86 runners documented in https://github.com/actions/runner-images/issues/9255 and https://github.com/actions/runner-images/issues/10686, which will impact our future upgrade behaviors.

I decided to also update the Xcode in the x86_64 runners, even though they are not being removed. It felt better to me to have all macOS runners on the same (major) version of Xcode. However, note that the x86_64 runners do not have the latest version of 15 (15.4), so I left them at 15.2 (which is currently the default Xcode of the runner).

Xcode 15 was previously causing problems (see #121058) which seem to be resolved now. @bjorn3 fixed the `invalid r_symbolnum` issue with cranelift. The issue with clang failing to link seems to be fixed, possibly by the update of the pre-built LLVM from 14 to llvm 15 in https://github.com/rust-lang/rust/pull/124850, or an update in our source version of LLVM. I have run some try builds and at least LLVM seems to build (I did not run any tests).

Closes #121058

ehuss committed:
(ci) Update macOS to Xcode 15

Mark-Simulacrum commented:
@bors r+ rollup=never

bors commented:
:pushpin: Commit 94579ff9301fcee893cd0c4ad9284ab5ef51e6cd has been approved by `Mark-Simulacrum`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Mark-Simulacrum 94579ff9301fcee893cd0c4ad9284ab5ef51e6cd -->
<!-- homu: {"type":"Approved","sha":"94579ff9301fcee893cd0c4ad9284ab5ef51e6cd","approver":"Mark-Simulacrum","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 94579ff9301fcee893cd0c4ad9284ab5ef51e6cd with merge 437be253971b531842a9dde04678e52229c39d86...
<!-- homu: {"type":"BuildStarted","head_sha":"94579ff9301fcee893cd0c4ad9284ab5ef51e6cd","merge_sha":"437be253971b531842a9dde04678e52229c39d86"} -->

rust-log-analyzer commented:
The job **`x86_64-apple-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/11375333771/job/31645817328) [(plain)](https://github.com/rust-lang-ci/rust/commit/437be253971b531842a9dde04678e52229c39d86/checks/31645817328/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

---- [debuginfo-lldb] tests/debuginfo/constant-ordering-prologue.rs stdout ----
NOTE: compiletest thinks it is using LLDB version 1500

error: check directive(s) from `/Users/runner/work/rust/rust/tests/debuginfo/constant-ordering-prologue.rs` not found in debugger output. errors:
    (constant-ordering-prologue.rs:23) ` = 19`
    (constant-ordering-prologue.rs:25) ` = 20`
    (constant-ordering-prologue.rs:27) ` = 21.5`
status: exit status: 0
command: PYTHONPATH="/Applications/Xcode_15.2.app/Contents/SharedFrameworks/LLDB.framework/Resources/Python" PYTHONUNBUFFERED="1" "/usr/bin/python3" "/Users/runner/work/rust/rust/src/etc/lldb_batchmode.py" "/Users/runner/work/rust/rust/build/x86_64-apple-darwin/test/debuginfo/constant-ordering-prologue.lldb/a" "/Users/runner/work/rust/rust/build/x86_64-apple-darwin/test/debuginfo/constant-ordering-prologue.lldb/constant-ordering-prologue.debugger.script"
LLDB batch-mode script
----------------------
----------------------
Debugger commands script is '/Users/runner/work/rust/rust/build/x86_64-apple-darwin/test/debuginfo/constant-ordering-prologue.lldb/constant-ordering-prologue.debugger.script'.
Target executable is '/Users/runner/work/rust/rust/build/x86_64-apple-darwin/test/debuginfo/constant-ordering-prologue.lldb/a'.
Current working directory is '/Users/runner/work/rust/rust'
Creating a target for '/Users/runner/work/rust/rust/build/x86_64-apple-darwin/test/debuginfo/constant-ordering-prologue.lldb/a'
settings set auto-confirm true
version
version
lldb-1500.0.200.58 Apple Swift version 5.9.2 (swiftlang-5.9.2.2.56 clang-1500.1.0.2.5) 
command script import /Users/runner/work/rust/rust/./src/etc/lldb_lookup.py
type synthetic add -l lldb_lookup.synthetic_lookup -x ".*" --category Rust
type summary add -F lldb_lookup.summary_lookup  -e -x -h "^(alloc::([a-z_]+::)+)String$" --category Rust
type summary add -F lldb_lookup.summary_lookup  -e -x -h "^&(mut )?str$" --category Rust
type summary add -F lldb_lookup.summary_lookup  -e -x -h "^&(mut )?\\[.+\\]$" --category Rust
type summary add -F lldb_lookup.summary_lookup  -e -x -h "^(std::ffi::([a-z_]+::)+)OsString$" --category Rust
type summary add -F lldb_lookup.summary_lookup  -e -x -h "^(alloc::([a-z_]+::)+)Vec<.+>$" --category Rust
type summary add -F lldb_lookup.summary_lookup  -e -x -h "^(alloc::([a-z_]+::)+)VecDeque<.+>$" --category Rust
type summary add -F lldb_lookup.summary_lookup  -e -x -h "^(alloc::([a-z_]+::)+)BTreeSet<.+>$" --category Rust
type summary add -F lldb_lookup.summary_lookup  -e -x -h "^(alloc::([a-z_]+::)+)BTreeMap<.+>$" --category Rust
type summary add -F lldb_lookup.summary_lookup  -e -x -h "^(std::collections::([a-z_]+::)+)HashMap<.+>$" --category Rust
type summary add -F lldb_lookup.summary_lookup  -e -x -h "^(std::collections::([a-z_]+::)+)HashSet<.+>$" --category Rust
type summary add -F lldb_lookup.summary_lookup  -e -x -h "^(alloc::([a-z_]+::)+)Rc<.+>$" --category Rust
type summary add -F lldb_lookup.summary_lookup  -e -x -h "^(alloc::([a-z_]+::)+)Arc<.+>$" --category Rust
type summary add -F lldb_lookup.summary_lookup  -e -x -h "^(core::([a-z_]+::)+)Cell<.+>$" --category Rust
type summary add -F lldb_lookup.summary_lookup  -e -x -h "^(core::([a-z_]+::)+)Ref<.+>$" --category Rust
type summary add -F lldb_lookup.summary_lookup  -e -x -h "^(core::([a-z_]+::)+)RefMut<.+>$" --category Rust
type summary add -F lldb_lookup.summary_lookup  -e -x -h "^(core::([a-z_]+::)+)RefCell<.+>$" --category Rust
type summary add -F lldb_lookup.summary_lookup  -e -x -h "^(core::([a-z_]+::)+)NonZero<.+>$" --category Rust
type summary add -F lldb_lookup.summary_lookup  -e -x -h "^core::num::([a-z_]+::)*NonZero.+$" --category Rust
type summary add -F lldb_lookup.summary_lookup  -e -x -h "^(std::([a-z_]+::)+)PathBuf$" --category Rust
type summary add -F lldb_lookup.summary_lookup  -e -x -h "^&(mut )?(std::([a-z_]+::)+)Path$" --category Rust
type category enable Rust
b "constant_ordering_prologue::binding"
DEBUG: breakpoint added, id = 1
DEBUG: breakpoint added, id = 1
Breakpoint 1: where = a`constant_ordering_prologue::binding::h6905e7563685df40 + 17 at constant-ordering-prologue.rs:30:9, address = 0x0000000100003eb1 
DEBUG: registering breakpoint callback, id = 1
Error while trying to register breakpoint callback, id = 1, message = error: could not get num args: can't find callable: breakpoint_callback
run
run
Process 16577 stopped * thread #1, name = 'main', queue = 'com.apple.main-thread', stop reason = breakpoint 1.1 frame #0: 0x0000000100003eb1 a`constant_ordering_prologue::binding::h6905e7563685df40(a=19, b=20, c=21.5) at constant-ordering-prologue.rs:30:9 27 // lldb-check: = 21.5 28 29 fn binding(a: i64, b: u64, c: f64) { -> 30 let x = 0; ^ 31 } 32 33 fn main() { Target 0: (a) stopped. Process 16577 launched: '/Users/runner/work/rust/rust/build/x86_64-apple-darwin/test/debuginfo/constant-ordering-prologue.lldb/a' (x86_64) 
(long) 19 
print b
(unsigned long) 20 
print c
```

</details>

jieyouxu commented:
```
2024-10-16T23:29:32.9094290Z failures:
2024-10-16T23:29:32.9095000Z     [debuginfo-lldb] tests/debuginfo/constant-ordering-prologue.rs
```

@bors r-

bjorn3 commented:
@bors r-

Bors got confused.

ehuss committed:
Update debuginfo test for newer lldb

For reasons I don't understand, lldb in Xcode 15 no longer prints objects as:

(long) $0 = 19

instead, it is printing them as:

(long) 19

ehuss commented:
I updated the test due to a small change in how lldb displays values in the newer xcode.

@bors r=Mark-Simulacrum

bors commented:
:pushpin: Commit 16b91e887afcdbfe0db2232639f0c9d33b7e2246 has been approved by `Mark-Simulacrum`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Mark-Simulacrum 16b91e887afcdbfe0db2232639f0c9d33b7e2246 -->
<!-- homu: {"type":"Approved","sha":"16b91e887afcdbfe0db2232639f0c9d33b7e2246","approver":"Mark-Simulacrum","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 16b91e887afcdbfe0db2232639f0c9d33b7e2246 with merge 4392847410ddd67f6734dd9845f9742ff9e85c83...
<!-- homu: {"type":"BuildStarted","head_sha":"16b91e887afcdbfe0db2232639f0c9d33b7e2246","merge_sha":"4392847410ddd67f6734dd9845f9742ff9e85c83"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/11447955844/job/31856528633)
Approved by: Mark-Simulacrum
Pushing 4392847410ddd67f6734dd9845f9742ff9e85c83 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"Mark-Simulacrum","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/11447955844/job/31856528633"},"merge_sha":"4392847410ddd67f6734dd9845f9742ff9e85c83"} -->

rust-timer commented:
Finished benchmarking commit (4392847410ddd67f6734dd9845f9742ff9e85c83): [comparison URL](https://perf.rust-lang.org/compare.html?start=31e102c509192189be841954f38962eb3cd9e2e2&end=4392847410ddd67f6734dd9845f9742ff9e85c83&stat=instructions:u).

### Overall result: ✅ improvements - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=31e102c509192189be841954f38962eb3cd9e2e2&end=4392847410ddd67f6734dd9845f9742ff9e85c83&stat=instructions:u)
This is the most reliable metric that we have; it was used to determine the overall result at the top of this comment. However, even this metric can sometimes exhibit noise.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -3.6% | [-3.6%, -3.6%] | 1     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -3.6% | [-3.6%, -3.6%] | 1     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=31e102c509192189be841954f38962eb3cd9e2e2&end=4392847410ddd67f6734dd9845f9742ff9e85c83&stat=max-rss)
<details>
<summary>Results (primary -2.3%, secondary -1.9%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -2.3% | [-2.3%, -2.3%] | 1     |
| Improvements ✅ <br /> (secondary) | -1.9% | [-2.8%, -1.1%] | 2     |
| All ❌✅ (primary)                 | -2.3% | [-2.3%, -2.3%] | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=31e102c509192189be841954f38962eb3cd9e2e2&end=4392847410ddd67f6734dd9845f9742ff9e85c83&stat=cycles:u)
<details>
<summary>Results (primary -3.9%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -3.9% | [-3.9%, -3.9%] | 1     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -3.9% | [-3.9%, -3.9%] | 1     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=31e102c509192189be841954f38962eb3cd9e2e2&end=4392847410ddd67f6734dd9845f9742ff9e85c83&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 782.781s -> 780.876s (-0.24%)
**Artifact size**: 333.65 MiB -> 333.66 MiB (0.00%)

cuviper commented:
Looks like this is **necessary** on beta as well, so I'm unilaterally backporting it.
(Xcode 14 failed in https://github.com/rust-lang/rust/pull/133077#issuecomment-2480076423)

@rustbot label +beta-accepted
